### PR TITLE
Fix a potential null pointer issue

### LINKED ISSUE #8249: TiFlash crash in LearnerReadWorker::waitUntilDataAvailable
```
0x7735a01\\tfaultSignalHandler(int, siginfo_t*, void*) [tiflash+125000193]\
                \\tlibs/libdaemon/src/BaseDaemon.cpp:214\
  0x7f78b25e7df0\\t<unknown symbol> [libc.so.6+347632]\
  0x7f78b2635b14\\t___pthread_mutex_lock [libc.so.6+666388]\
  0x7f78b7185f46\\tstd::__1::mutex::lock() [libc++.so.1+421702]\
       0x8afcfdc\\tDB::LearnerReadWorker::waitUntilDataAvailable(std::__1::unordered_map<unsigned long, DB::RegionLearnerReadSnapshot, std::__1::hash<unsigned long>, std::__1::equal_to<unsigned long>, std::__1::allocator<std::__1::pair<unsigned long const, DB::RegionLearnerReadSnapshot> > > const&, unsigned long, unsigned long) [tiflash+145739740]\
                \\tdbms/src/Storages/KVStore/Read/LearnerReadWorker.cpp:415\
       0x8af7603\\tDB::doLearnerRead(long, DB::MvccQueryInfo&, bool, DB::Context&, std::__1::shared_ptr<DB::Logger> const&) [tiflash+145716739]\             
                \\tdbms/src/Storages/KVStore/Read/LearnerRead.cpp:57\
       0x87936c7\\tDB::DAGStorageInterpreter::prepare() [tiflash+142161607]\
                \\tdbms/src/Flash/Coprocessor/DAGStorageInterpreter.cpp:635\
       0x8999d17\\tDB::PhysicalTableScan::buildPipeline(DB::PipelineBuilder&, DB::Context&, DB::PipelineExecutorContext&) [tiflash+144284951]\  
                \\tdbms/src/Flash/Planner/Plans/PhysicalTableScan.cpp:137\
       0x8925628\\tDB::PhysicalAggregation::buildPipeline(DB::PipelineBuilder&, DB::Context&, DB::PipelineExecutorContext&) [tiflash+143808040]\
                \\tdbms/src/Flash/Planner/Plans/PhysicalAggregation.cpp:278\
       0x8915df2\\tDB::PhysicalPlanNode::buildPipeline(DB::PipelineBuilder&, DB::Context&, DB::PipelineExecutorContext&) [tiflash+143744498]\
                \\tdbms/src/Flash/Planner/PhysicalPlanNode.cpp:116\         
       0x8915df2\\tDB::PhysicalPlanNode::buildPipeline(DB::PipelineBuilder&, DB::Context&, DB::PipelineExecutorContext&) [tiflash+143744498]\
                \\tdbms/src/Flash/Planner/PhysicalPlanNode.cpp:116\
       0x891133d\\tDB::PhysicalPlan::toPipeline(DB::PipelineExecutorContext&, DB::Context&) [tiflash+143725373]\
                \\tdbms/src/Flash/Planner/PhysicalPlan.cpp:325\
       0x88c97b7\\tDB::PipelineExecutor::PipelineExecutor(std::__1::shared_ptr<MemoryTracker> const&, DB::AutoSpillTrigger*, std::__1::function<void (std::__1::shared_ptr<DB::OperatorSpillContext> const&)> const&, DB::Context&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) [tiflash+143431607]\
                \\tdbms/src/Flash/Executor/PipelineExecutor.cpp:45\ 
       0x8722e80\\tDB::(anonymous namespace)::executeAsPipeline(DB::Context&, bool) [tiflash+141700736]\
                \\tdbms/src/Flash/executeQuery.cpp:198\
       0x8722514\\tDB::queryExecute(DB::Context&, bool) [tiflash+141698324]\
                \\tdbms/src/Flash/executeQuery.cpp:238\
       0x886b043\\tDB::MPPTask::runImpl() [tiflash+143044675]\
                \\tdbms/src/Flash/Mpp/MPPTask.cpp:527\
       0x1fb5488\\tauto DB::wrapInvocable<std::__1::function<void ()> >(bool, std::__1::function<void ()>&&)::'lambda'()::operator()() [tiflash+33248392]\
                \\tdbms/src/Common/wrapInvocable.h:36\
       0x1ea7953\\tDB::DynamicThreadPool::executeTask(std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >&) [tiflash+32143699]\
                \\tdbms/src/Common/DynamicThreadPool.cpp:124\
       0x1ea72e1\\tDB::DynamicThreadPool::fixedWork(unsigned long) [tiflash+32142049]\
                \\tdbms/src/Common/DynamicThreadPool.cpp:138\
       0x1ea8fc2\\tvoid* std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, std::__1::thread DB::ThreadFactory::newThread<void (DB::DynamicThreadPool::*)(unsigned long), DB::DynamicThreadPool*, unsigned long&>(bool, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, void (DB::DynamicThreadPool::*&&)(unsigned long), DB::DynamicThreadPool*&&, unsigned long&)::'lambda'(auto&&...), DB::DynamicThreadPool*, unsigned long> >(void*) [tiflash+32149442]\
```

```
tc-tiflash-0/server.log
916353:[2023/10/26 03:43:30.370 +08:00] [ERROR] [BaseDaemon.cpp:370] [########################################] [source=BaseDaemon] [thread_id=5289]
916360:[2023/10/26 03:43:30.370 +08:00] [ERROR] [BaseDaemon.cpp:371] ["(from thread 1251) Received signal Segmentation fault(11)."] [source=BaseDaemon] [thread_id=5289]
916361:[2023/10/26 03:43:30.370 +08:00] [ERROR] [BaseDaemon.cpp:401] ["Address: 0x278"] [source=BaseDaemon] [thread_id=5289]
916362:[2023/10/26 03:43:30.370 +08:00] [ERROR] [BaseDaemon.cpp:407] ["Access: read."] [source=BaseDaemon] [thread_id=5289]
916369:[2023/10/26 03:43:30.370 +08:00] [ERROR] [BaseDaemon.cpp:416] ["Address not mapped to object."] [source=BaseDaemon] [thread_id=5289]
916883:[2023/10/26 03:43:30.396 +08:00] [ERROR] [BaseDaemon.cpp:563] ["\n       0x7735a01\tfaultSignalHandler(int, siginfo_t*, void*) [tiflash+125000193]\n                \tlibs/libdaemon/src/BaseDaemon.cpp:214\n  0x7f78b25e7df0\t<unknown symbol> [libc.so.6+347632]\n  0x7f78b2635b14\t___pthread_mutex_lock [libc.so.6+666388]\n  0x7f78b7185f46\tstd::__1::mutex::lock() [libc++.so.1+421702]\n       0x8afcfdc\tDB::LearnerReadWorker::waitUntilDataAvailable(std::__1::unordered_map<unsigned long, DB::RegionLearnerReadSnapshot, std::__1::hash<unsigned long>, std::__1::equal_to<unsigned long>, std::__1::allocator<std::__1::pair<unsigned long const, DB::RegionLearnerReadSnapshot> > > const&, unsigned long, unsigned long) [tiflash+145739740]\n                \tdbms/src/Storages/KVStore/Read/LearnerReadWorker.cpp:415\n       0x8af7603\tDB::doLearnerRead(long, DB::MvccQueryInfo&, bool, DB::Context&, std::__1::shared_ptr<DB::Logger> const&) [tiflash+145716739]\n                \tdbms/src/Storages/KVStore/Read/LearnerRead.cpp:57\n       0x87936c7\tDB::DAGStorageInterpreter::prepare() [tiflash+142161607]\n                \tdbms/src/Flash/Coprocessor/DAGStorageInterpreter.cpp:635\n       0x8999d17\tDB::PhysicalTableScan::buildPipeline(DB::PipelineBuilder&, DB::Context&, DB::PipelineExecutorContext&) [tiflash+144284951]\n                \tdbms/src/Flash/Planner/Plans/PhysicalTableScan.cpp:137\n       0x8925628\tDB::PhysicalAggregation::buildPipeline(DB::PipelineBuilder&, DB::Context&, DB::PipelineExecutorContext&) [tiflash+143808040]\n                \tdbms/src/Flash/Planner/Plans/PhysicalAggregation.cpp:278\n       0x8915df2\tDB::PhysicalPlanNode::buildPipeline(DB::PipelineBuilder&, DB::Context&, DB::PipelineExecutorContext&) [tiflash+143744498]\n                \tdbms/src/Flash/Planner/PhysicalPlanNode.cpp:116\n       0x8915df2\tDB::PhysicalPlanNode::buildPipeline(DB::PipelineBuilder&, DB::Context&, DB::PipelineExecutorContext&) [tiflash+143744498]\n                \tdbms/src/Flash/Planner/PhysicalPlanNode.cpp:116\n       0x891133d\tDB::PhysicalPlan::toPipeline(DB::PipelineExecutorContext&, DB::Context&) [tiflash+143725373]\n                \tdbms/src/Flash/Planner/PhysicalPlan.cpp:325\n       0x88c97b7\tDB::PipelineExecutor::PipelineExecutor(std::__1::shared_ptr<MemoryTracker> const&, DB::AutoSpillTrigger*, std::__1::function<void (std::__1::shared_ptr<DB::OperatorSpillContext> const&)> const&, DB::Context&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) [tiflash+143431607]\n                \tdbms/src/Flash/Executor/PipelineExecutor.cpp:45\n       0x8722e80\tDB::(anonymous namespace)::executeAsPipeline(DB::Context&, bool) [tiflash+141700736]\n                \tdbms/src/Flash/executeQuery.cpp:198\n       0x8722514\tDB::queryExecute(DB::Context&, bool) [tiflash+141698324]\n                \tdbms/src/Flash/executeQuery.cpp:238\n       0x886b043\tDB::MPPTask::runImpl() [tiflash+143044675]\n                \tdbms/src/Flash/Mpp/MPPTask.cpp:527\n       0x1fb5488\tauto DB::wrapInvocable<std::__1::function<void ()> >(bool, std::__1::function<void ()>&&)::'lambda'()::operator()() [tiflash+33248392]\n                \tdbms/src/Common/wrapInvocable.h:36\n       0x1ea7953\tDB::DynamicThreadPool::executeTask(std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >&) [tiflash+32143699]\n                \tdbms/src/Common/DynamicThreadPool.cpp:124\n       0x1ea72e1\tDB::DynamicThreadPool::fixedWork(unsigned long) [tiflash+32142049]\n                \tdbms/src/Common/DynamicThreadPool.cpp:138\n       0x1ea8fc2\tvoid* std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, std::__1::thread DB::ThreadFactory::newThread<void (DB::DynamicThreadPool::*)(unsigned long), DB::DynamicThreadPool*, unsigned long&>(bool, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, void (DB::DynamicThreadPool::*&&)(unsigned long), DB::DynamicThreadPool*&&, unsigned long&)::'lambda'(auto&&...), DB::DynamicThreadPool*, unsigned long> >(void*) [tiflash+32149442]\n                \t/usr/local/bin/../include/c++/v1/thread:291\n  0x7f78b2632802\tstart_thread [libc.so.6+653314]"] [source=BaseDaemon] [thread_id=5289]
```

```
tc-tiflash-2/server.log.2023-10-26-03:45:13.358
1684161:[2023/10/26 03:43:11.789 +08:00] [ERROR] [BaseDaemon.cpp:370] [########################################] [source=BaseDaemon] [thread_id=6680]
1684164:[2023/10/26 03:43:11.789 +08:00] [ERROR] [BaseDaemon.cpp:371] ["(from thread 711) Received signal Segmentation fault(11)."] [source=BaseDaemon] [thread_id=6680]
1684165:[2023/10/26 03:43:11.789 +08:00] [ERROR] [BaseDaemon.cpp:401] ["Address: 0x278"] [source=BaseDaemon] [thread_id=6680]
1684166:[2023/10/26 03:43:11.789 +08:00] [ERROR] [BaseDaemon.cpp:407] ["Access: read."] [source=BaseDaemon] [thread_id=6680]
1684197:[2023/10/26 03:43:11.790 +08:00] [ERROR] [BaseDaemon.cpp:416] ["Address not mapped to object."] [source=BaseDaemon] [thread_id=6680]
1684416:[2023/10/26 03:43:11.827 +08:00] [ERROR] [BaseDaemon.cpp:563] ["\n       0x7733431\tfaultSignalHandler(int, siginfo_t*, void*) [tiflash+124990513]\n                \tlibs/libdaemon/src/BaseDaemon.cpp:214\n  0x7fd40f3cfdf0\t<unknown symbol> [libc.so.6+347632]\n  0x7fd40f41db14\t___pthread_mutex_lock [libc.so.6+666388]\n  0x7fd413f6df46\tstd::__1::mutex::lock() [libc++.so.1+421702]\n       0x8aff65c\tDB::LearnerReadWorker::waitUntilDataAvailable(std::__1::unordered_map<unsigned long, DB::RegionLearnerReadSnapshot, std::__1::hash<unsigned long>, std::__1::equal_to<unsigned long>, std::__1::allocator<std::__1::pair<unsigned long const, DB::RegionLearnerReadSnapshot> > > const&, unsigned long, unsigned long) [tiflash+145749596]\n                \tdbms/src/Storages/KVStore/Read/LearnerReadWorker.cpp:415\n       0x8af9c83\tDB::doLearnerRead(long, DB::MvccQueryInfo&, bool, DB::Context&, std::__1::shared_ptr<DB::Logger> const&) [tiflash+145726595]\n                \tdbms/src/Storages/KVStore/Read/LearnerRead.cpp:57\n       0x8795997\tDB::DAGStorageInterpreter::prepare() [tiflash+142170519]\n                \tdbms/src/Flash/Coprocessor/DAGStorageInterpreter.cpp:635\n       0x899c107\tDB::PhysicalTableScan::buildPipeline(DB::PipelineBuilder&, DB::Context&, DB::PipelineExecutorContext&) [tiflash+144294151]\n                \tdbms/src/Flash/Planner/Plans/PhysicalTableScan.cpp:137\n       0x8927a18\tDB::PhysicalAggregation::buildPipeline(DB::PipelineBuilder&, DB::Context&, DB::PipelineExecutorContext&) [tiflash+143817240]\n                \tdbms/src/Flash/Planner/Plans/PhysicalAggregation.cpp:278\n       0x89181e2\tDB::PhysicalPlanNode::buildPipeline(DB::PipelineBuilder&, DB::Context&, DB::PipelineExecutorContext&) [tiflash+143753698]\n                \tdbms/src/Flash/Planner/PhysicalPlanNode.cpp:116\n       0x89181e2\tDB::PhysicalPlanNode::buildPipeline(DB::PipelineBuilder&, DB::Context&, DB::PipelineExecutorContext&) [tiflash+143753698]\n                \tdbms/src/Flash/Planner/PhysicalPlanNode.cpp:116\n       0x891372d\tDB::PhysicalPlan::toPipeline(DB::PipelineExecutorContext&, DB::Context&) [tiflash+143734573]\n                \tdbms/src/Flash/Planner/PhysicalPlan.cpp:325\n       0x88cbba7\tDB::PipelineExecutor::PipelineExecutor(std::__1::shared_ptr<MemoryTracker> const&, DB::AutoSpillTrigger*, std::__1::function<void (std::__1::shared_ptr<DB::OperatorSpillContext> const&)> const&, DB::Context&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) [tiflash+143440807]\n                \tdbms/src/Flash/Executor/PipelineExecutor.cpp:45\n       0x8725190\tDB::(anonymous namespace)::executeAsPipeline(DB::Context&, bool) [tiflash+141709712]\n                \tdbms/src/Flash/executeQuery.cpp:198\n       0x8724824\tDB::queryExecute(DB::Context&, bool) [tiflash+141707300]\n                \tdbms/src/Flash/executeQuery.cpp:238\n       0x886d013\tDB::MPPTask::runImpl() [tiflash+143052819]\n                \tdbms/src/Flash/Mpp/MPPTask.cpp:527\n       0x1fb5458\tauto DB::wrapInvocable<std::__1::function<void ()> >(bool, std::__1::function<void ()>&&)::'lambda'()::operator()() [tiflash+33248344]\n                \tdbms/src/Common/wrapInvocable.h:36\n       0x1ea7dc3\tDB::DynamicThreadPool::executeTask(std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >&) [tiflash+32144835]\n                \tdbms/src/Common/DynamicThreadPool.cpp:124\n       0x1ea7751\tDB::DynamicThreadPool::fixedWork(unsigned long) [tiflash+32143185]\n                \tdbms/src/Common/DynamicThreadPool.cpp:138\n       0x1ea9432\tvoid* std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, std::__1::thread DB::ThreadFactory::newThread<void (DB::DynamicThreadPool::*)(unsigned long), DB::DynamicThreadPool*, unsigned long&>(bool, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, void (DB::DynamicThreadPool::*&&)(unsigned long), DB::DynamicThreadPool*&&, unsigned long&)::'lambda'(auto&&...), DB::DynamicThreadPool*, unsigned long> >(void*) [tiflash+32150578]\n                \t/usr/local/bin/../include/c++/v1/thread:291\n  0x7fd40f41a802\tstart_thread [libc.so.6+653314]"] [source=BaseDaemon] [thread_id=6680]
```

### PR DESCRIPTION
**Pull request by CalvinNeo**
### What problem does this PR solve?

Issue Number: close #8249

Problem Summary:

This may not be the root cause, but it can panic.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Storages/KVStore/Read/LearnerReadWorker.cpp [hunk 1]
```diff
@@ -202,11 +202,19 @@ void LearnerReadWorker::recordReadIndexError(RegionsReadIndexResult & read_index
...
+                    extra_msg = fmt::format(
+                        "read_index_resp error, region_id={} version={} conf_version={}",
+                        region_id,
+                        kv_region->version(),
+                        kv_region->confVer());
+                }
+                else
+                {
+                    extra_msg = fmt::format("read_index_resp error, region_id={} does not exist", region_id);
+                }
```
Comment (JaySon-Huang) [inline dbms/src/Storages/KVStore/Read/LearnerReadWorker.cpp:217]: I think we should use the region in `regions_snapshot` to get the `version` and `confVersion` to the logging message?

Comment (JaySon-Huang) [reply dbms/src/Storages/KVStore/Read/LearnerReadWorker.cpp:217]: or get it directly from the protobuf struct? https://github.com/pingcap/kvproto/blob/c0689aded40cd69294399a7603544fc5336cc845/proto/errorpb.proto#L81-L87

Comment (CalvinNeo) [reply dbms/src/Storages/KVStore/Read/LearnerReadWorker.cpp:217]: The information in `regions_snapshot` will be printed in an outer exception. However, we can still add it here.

Comment (JaySon-Huang) [reply dbms/src/Storages/KVStore/Read/LearnerReadWorker.cpp:217]: We build the read index according to the region meta in `regions_snapshot`. The region meta in `kvstore` could be updated during the learner read.
Which message do we want to show in the error?

Comment (CalvinNeo) [reply dbms/src/Storages/KVStore/Read/LearnerReadWorker.cpp:217]: In doCopLearnerRead, the region info from cop differs from the region info from KVStore. The KVStore has the most up-to-date region epoch, so there is a very large chance that the region info from cop is obsolete, we add log to check that. See https://github.com/pingcap/tiflash/issues/8253

**Review by JaySon-Huang — APPROVED** [review]
LGTM

**Comment by CalvinNeo** [comment]
Not stable test
```

[2023-10-27T08:46:48.364Z] fullstack-test2/ddl/alter_partition_by.test: Running

[2023-10-27T08:46:58.284Z]   File: fullstack-test2/ddl/alter_partition_by.test

[2023-10-27T08:46:58.284Z]   Error line: 171

[2023-10-27T08:46:58.284Z]   Error: select tidb_database,tidb_name from system.tables where tidb_database = 'test' and tidb_name = 't2' and is_tombstone = 0

[2023-10-27T08:46:58.284Z]   Result:

[2023-10-27T08:46:58.284Z]     <nothing>

[2023-10-27T08:46:58.284Z]   Expected:

[2023-10-27T08:46:58.284Z]     ┌─tidb_database─┬─tidb_name─┐

[2023-10-27T08:46:58.284Z]     │ test          │ t2        │

[2023-10-27T08:46:58.284Z]     └───────────────┴───────────┘
```

**Pull request merged**